Source/Charts/Data/Implementations/Standard/ChartDataSet.swift

```diff
@@ -517,3 +480,37 @@ open class ChartDataSet: ChartBaseDataSet
         return copy
     }
 }
+
+extension RandomAccessCollection {
+    func sortedInsertionPoint<T>(of value: T, sortedAscending: (T, Element) -> Bool) -> Index {
+        var slice = self[...]
```

Comment:
I bet, using array slices here is not required. Yes, array is not copied when slice is created, but array slice still seems to be more complex, than just array index

Reply:
@anton-filimonov 
Answering everything else here


This is what I started with, in which case absolutely slices make sense:
```
extension RandomAccessCollection where Element: Comparable {
    func sortedInsertionPoint(of value: Element) -> Index {
        if isEmpty { return startIndex }
        let middle = index(startIndex, offsetBy: count / 2)
        if value < self[middle] {
            return self[..<middle].sortedInsertionPoint(of: value)
        } else {
            return self[index(after: middle)...].sortedInsertionPoint(of: value)
        }
    }
}
```
However, there tends to be less optimization for recursive functions. Turning that into a loop which is optimized becomes:
```
extension RandomAccessCollection where Element: Comparable {
    func sortedInsertionPoint(of value: Element) -> Index {
        var slice = self[...]

        while !slice.isEmpty {
            let middle = slice.index(slice.startIndex, offsetBy: slice.count / 2)
            if value < slice[middle] {
                slice = slice[..<middle]
            } else {
                slice = slice[index(after: middle)...]
            }
        }

        return slice.startIndex
    }
}
```
I think this still reads nicely. It becomes weird however, when we're trying to compare a value to a property of `Element` which is why I introduced the version of the function with the closure.

As for the `sortedInsertionPoint` name, I think it makes sense when reading it as "What index in the collection should this new element be inserted at", but yes, it probably could be better. Note that it is not exposed in the API. Sorted `Collection`s and binary search in Swift are summed up [here](https://forums.swift.org/t/review-se-0074-implementation-of-binary-search-functions/2438/5?u=jjatie) and [here](https://forums.swift.org/t/sorted-collection-protocol/11738) I'm happy to bike shed some names for this.

Ultimately, `ChartDataSet` should conform to `RandomAccessCollection`, and there should be two versions of this method. One for internal use that returns a slice to optimize our internal implementations, and a public one that returns a `ChartDataSet` or `Array<ChartDataEntry>`.